nixos/modules/services/web-apps/akkoma.nix

```diff
@@ -0,0 +1,1031 @@
+    ''];
+
+    users = {
+      users."${cfg.user}" = {
+        description = "Akkoma user";
+        group = cfg.group;
+        isSystemUser = true;
+      };
+      groups."${cfg.group}" = { };
+    };
+
+    # Confinement of the main service unit requires separation of the
+    # configuration generation into a separate unit to permit access to secrets
+    # residing outside of the chroot.
+    systemd.services.akkoma-config = {
+      description = "Akkoma social network configuration";
+      reloadTriggers = [ configFile ] ++ secretPaths;
+
+      unitConfig.PropagatesReloadTo = [ "akkoma.service" ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        UMask = "0077";
+
+        RuntimeDirectory = "akkoma";
+        RuntimeDirectoryMode = "0711";
+
+        ExecStart =
+          (if cfg.dist.cookie == null
+            then [ genScript ]
+            else [ copyScript ])
+          ++ optionals cfg.initSecrets [ initSecretsScript ]
+          ++ [ configScript ];
+        ExecReload = optionals cfg.initSecrets [ initSecretsScript ]
+          ++ [ configScript ];
+      };
+    };
+
+    systemd.services.akkoma-initdb = mkIf cfg.initDb.enable {
+      description = "Akkoma social network database setup";
+      requires = [ "akkoma-config.service" ];
+      requiredBy = [ "akkoma.service" ];
+      after = [ "akkoma-config.service" "postgresql.service" ];
+      before = [ "akkoma.service" ];
+
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        UMask = "0077";
+        ExecStart = initDbScript;
+        PrivateTmp = true;
+      } // optionalAttrs (db ? socket_dir || db ? socket) {
+        User = cfg.initDb.username;
+      };
+    };
+
+    systemd.services.akkoma = let
+      runtimeInputs = with pkgs; [ coreutils gawk gnused ] ++ cfg.extraPackages;
+    in {
+      description = "Akkoma social network";
+      documentation = [ "https://docs.akkoma.dev/stable/" ];
+
+      # This service depends on network-online.target and is sequenced after
+      # it because it requires access to the Internet to function properly.
+      bindsTo = [ "akkoma-config.service" ];
+      wants = [ "network-online.service" ];
+      wantedBy = [ "multi-user.target" ];
+      after = [
+        "akkoma-config.target"
+        "network.target"
+        "network-online.target"
+        "postgresql.service"
+      ];
+
+      confinement.packages = runtimeInputs;
+      path = runtimeInputs;
+
+      serviceConfig = {
+        Type = "exec";
+        User = cfg.user;
+        Group = cfg.group;
+        UMask = "0077";
+
+        RuntimeDirectory = "akkoma";
+        RuntimeDirectoryMode = "0711";
+        RuntimeDirectoryPreserve = true;
+        StateDirectory = "akkoma akkoma/static akkoma/uploads";
+        StateDirectoryMode = "0700";
+
+        BindReadOnlyPaths = [ "/etc/hosts" "/etc/resolv.conf" ]
```

Comment:
Could you help me understand which other option causes the whole filesystem (with the exception of `/home`) _not_ to be present? Is it just the presence of this option?

Reply:
These options on their own do not do that.

Except for the frontend files, these are essentially no‐ops if `confinement.enable` is not enabled.

Even though there is an automated test that runs with confinement enabled I was asked to not enable it by default.